Add whitespace stripping in the input

Kijewski opened the pull request:
In HTML sources indentation can aid readability, but it only bloats the
output. This PR adds the argument "strip" to templates, which lets you
automatically strip the template input:

* "none": No stripping of the input. The generator still strips spaces
  at the back of the input.
* ~~"tail": The current default, i.e. pop a single newline character at
  the end.~~
* ~~"trim": remove all spaces at the front and back of a line, and remove
  empty lines.~~
* "eager": like "trim", but replace multiple whitespaces with a single
  space.

The automatic stripping does not try to understand the input at all. It
does not know what `xml:space="preserve"`, `<pre>` or `<textarea>`
means.

vallentin commented:
I "don't" like the idea of `trim`. Because whitespace is or can be significant in HTML/CSS, i.e. with inline menus :)

Also, can this be used for #581? Because I looked a bit at that before Christmas, but then I got busy with Christmas and New Year's

Kijewski commented:
It's like trim() is called on every line, then empty lines are dropped, and for the eager case `s/\s+/ /g` applied.

Of course, white spaces are significant in `<pre>` and `<textarea>`. But how often do you see `<pre>` nowadays? When do you use a `<textarea>` that is populated by static data, not by `{{data}}`? For these two cases you cannot apply the automatic stripping.

djc commented on askama_shared/src/lib.rs:
```diff
@@ -257,19 +258,71 @@ where
+    None,
+    /// Remove a single single newline at the end of the input. This is the default.
+    Tail,
+    /// Remove all whitespaces at the front and back all lines, and remove empty lines.
```
Missing `of`? Please remove periods at the end of the first line of docstrings.

Kijewski replied:
Fixed.

djc commented on askama_shared/src/lib.rs:
```diff
@@ -257,19 +258,71 @@ where
+    /// Remove a single single newline at the end of the input. This is the default.
+    Tail,
+    /// Remove all whitespaces at the front and back all lines, and remove empty lines.
+    Trim,
```
Maybe call this `TrimLines` to avoid confusion as with @vallentin?

Kijewski replied:
I renamed it into TrimLines / "trim-lines".

djc commented on askama_shared/src/lib.rs:
```diff
@@ -257,19 +258,71 @@ where
+
+impl Strip {
+    pub fn apply<'a, S: Into<Cow<'a, str>>>(self, src: S) -> String {
+        let src = S::into(src);
```
I think I'd prefer spelling this `src.into()`.

Kijewski replied:
You're right. I don't know why I used the more complicated syntax before. :)

djc reviewed (commented):
Awesome that you're working on this, did you guys already see https://github.com/rust-lang/rust/pull/92526? It's pretty amazing! This also came up there. I'll link to this PR in from there so that we can feedback from them.

I guess this also needs some good documentation in the book.

My main point of feedback is that I think we'd want a way to change this setting on a crate-wide basis, in the `askama.toml` configuration. There, we might make it part of an extension/type-based configuration (I think we already have some?) so that people can easily apply this to HTML but not plain text templates. I was going to suggest that we ditch any kind of per-template configuration but it just occurs to me that that might make it harder to test this well (end-to-end) the way you've done here.

camelid commented:
Am I correct in thinking that this is all done at compile-time? If so, that makes this extra-exciting :)

djc commented:
> Am I correct in thinking that this is all done at compile-time? If so, that makes this extra-exciting :)

Yes, that would all be done at compile time.

Kijewski commented:
I moved Strip from askama_shared's root into askama_shared::input. That's were the other arguments life.

djc commented:
Did you have any thoughts about the configuration approach?

Kijewski commented:
> Did you have any thoughts about the configuration approach?

Yes, that should be a project-wide configuration. I added the optional strip parameter to the config, too.

But I actually have not idea how to write a test for it. Can you give me some pointers?

Kijewski commented:
I'm not convinced anymore that this is the best implementation.

Most likely it would be simpler and better to use Strip::apply() in Generator::write_buf_writable() instead of preformatting the input.

Kijewski commented:
I simplified the PR. No only "none" and "eager" are allowed.

Kijewski commented:
My "simplified" version does not work. The "complicated" version https://github.com/djc/askama/commit/766d049536c17a089fd7031e639d9f4999c9e64c did. Reverted and rebased.

Kijewski commented:
Superseded by #673.
